include/openthread/thread.h

```diff
@@ -380,14 +381,36 @@ otError otThreadSetLinkMode(otInstance *aInstance, otLinkModeConfig aConfig);
 /**
  * Get the thrNetworkKey.
  *
+ * This function will return an union of key reference and literal key. These feilds
+ * will be populated based on the Crypto module being used. For PSA Crypto, a reference
+ * to the thrNetworkKey is returned and not the literal key.
+ *
  * @param[in]   aInstance   A pointer to an OpenThread instance.
+ * @param[out]  aKeyType    A output pointer to store the key type
  *
  * @returns A pointer to a buffer containing the thrNetworkKey.
  *
- * @sa otThreadSetNetworkKey
+ * @sa otThreadSetNetworkKey otThreadCopyNetworkKey
+ *
+ */
+const otNetworkKey *otThreadGetNetworkKey(otInstance *aInstance, otCryptoType *aKeyType);
+
+/**
+ * Copy the thrNetworkKey into passed buffer
+ *
+ * This function copies the thrNetworkKey as a literal into the output buffer and works
+ * irrespective of the Crypto module used.
+ *
+ * @param[in]   aInstance   A pointer to an OpenThread instance.
+ * @param[out]  aNetworkKey A pointer to thrNetworkKey buffer.
  *
+ * @retval OT_ERROR_NONE            Successfully copied the thrNetworkKey.
+ * @retval OT_ERROR_NOT_SUPPORTED   Operation was not supported due to key export restrictions.
+ * @retval OT_ERROR_FAILED          Copying the thrNetworkKey failed.
+ *
+ * @sa otThreadSetNetworkKey otThreadCopyNetworkKey
  */
-const otNetworkKey *otThreadGetNetworkKey(otInstance *aInstance);
+otError otThreadCopyNetworkKey(otInstance *aInstance, otNetworkKey *aNetworkKey);
```

Comment:
Do we need the `CopyNetworkKey` public APIs? 
We already provide `GetNetworkKey` and caller can choose to copy it into another buffer if they want to.

Reply:
I would prefer to change the current implementation `GetNetworkKey` with the new implementation in  `CopyNetworkKey`. This is because when the user gets the network key he either has the literal key or key ref. If it is key Ref the user cannot invoke the `CopyKey` from the `NetworkKey` class, as he does not have access to it. I would prefer if the user passes a buffer to `GetNetworkKey`  and we call CopyKey from `GetNetworkKey`. This way user does not need to know anything about key Refs or literal keys, the `GetNetworkKey` will abstract this.

Reply:
If given crypto engine is configured in a way that user can export the key value, key reference should be enough for the user. User could call something like `psa_export_key()` to get the key value. If crypto engine is configured to prevent user from accessing key values, OpenThread should not work around this rule.

I think `GetNetworkKey` should be enough and cleaner API.

Reply:
But that would mean the same change will go in cli code as well. Now the cli.cpp needs to call the `psa_export_key` to get the key literals. 

My proposal is to pass the buffer as an output pointer to `GetNetworkKey` This allows us to return `otError` to caller to indicate failures. So if the crypto engine is configured to prevent the access to the key this API will return error.

Reply:
API user may want to use key ref retrieved from OT. It might be allowed by crypto configuration to perform crypto operations by API user, but not to export the key. That's why I think user should get key reference if such model is in use. API user (like CLI) can try to export key if it needs its value.

I'm not sure if CLI is considered internal OpenThread module and if it can use `patform/crypto.h` to make key exporting easier.

Reply:
Yes, that does make sense. We need a resolution on using spa APIs in cli.cpp. If it is OK to be used in cli, I will revert the implementation to use `GetNetworkKey`. If not I will add a helper API to export the key for any given keyId.

Reply:
@abtink @hubertmis @LuDuda @Damian-Nordic @jwhui Taking Abtin's suggestion I am creating a new PR with all the changes and closing this one. 

Please refer to https://github.com/openthread/openthread/pull/6862

The new PR has been updated and contains all the changes we discussed.